### PR TITLE
feat(sql): now() function in select queries

### PR DESCRIPTION
**Pull request by ideoma**

### CODE REVIEW

File: core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java [hunk 1]
```diff
@@ -51,9 +54,9 @@
     private int initialPartitionLo;
     private int initialPartitionHi;
 
-    public AbstractIntervalDataFrameCursor(@Transient LongList intervals, int timestampIndex) {
+    public AbstractIntervalDataFrameCursor(@Transient RuntimeIntrinsicIntervalModel intervals, int timestampIndex) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java:57]: remove transient annotation

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/questdb/griffin/engine/functions/BinaryFunction.java [hunk 2]
```diff
@@ -56,4 +56,8 @@ default void toTop() {
     Function getLeft();
 
     Function getRight();
+
+    default boolean isRuntimeConstant() {
+        return (getLeft().isRuntimeConstant() || getLeft().isConstant()) && (getRight().isConstant() || getRight().isRuntimeConstant());
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/BinaryFunction.java:61]: i don't feel this is correct, perhaps this:

```java
        if ((!getLeft().isConstant() && !getLeft().isRuntimeConstant()) || (!getRight().isConstant() && !getRight().isRuntimeConstant())) {
            return false;
        }
        return getLeft().isRuntimeConstant() || getRight().isRuntimeConstant();

```

collapsible to 

```java
        final Function l = getLeft();
        final Function r = getRight();
        return (l.isConstant() || l.isRuntimeConstant()) && (l.isConstant() || l.isRuntimeConstant()) && (l.isRuntimeConstant() || r.isRuntimeConstant());
```

File: core/src/main/java/io/questdb/griffin/engine/functions/math/SubTimestampFunctionFactory.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+        public Func(int position, Function left, Function right) {
+            super(position);
+            this.left = left;
+            this.right = right;
+        }
+
+        @Override
+        public long getTimestamp(Record rec) {
+            return left.getTimestamp(rec) - right.getTimestamp(rec);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/math/SubTimestampFunctionFactory.java:59]: does not handle nulls

File: core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModel.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            }
+        }
+        return true;
+    }
+
+    private void addEvaluateDynamicIntervals(LongList outIntervals, SqlExecutionContext sqlContext) {
+        int dynamicStart = intervals.size() - dynamicRangeList.size() * STATIC_LONGS_PER_DYNAMIC_INTERVAL;
+        int dynamicIndex = 0;
+
+        for (int i = dynamicStart; i < intervals.size(); i += STATIC_LONGS_PER_DYNAMIC_INTERVAL) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModel.java:92]: could be slightly faster loop with pre-calculated size

File: core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModel.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+            Function dynamicFunction = dynamicRangeList.get(dynamicIndex++);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModel.java:93]: pls use `getQuick()`

File: core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModel.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+                if (operation == IntervalOperation.INTERSECT_EQUALS) {
+                    hi = lo = dynamicValue;
+                } else if (hi == DYNAMIC_LO_HI) {
+                    hi = dynamicValue + adjustment;
+                } else {
+                    lo = dynamicValue + adjustment;
+                }
+
+                outIntervals.extendAndSet(divider + 1, hi);
+                outIntervals.extendAndSet(divider, lo);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModel.java:122]: the latter can be `setQuick`

File: core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModelBuilder.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+ * When first interval involving function is added all data starts to be encoded in 4 longs in staticPeriods
+ * 0: lo (long)
+ * 1: hi (long)
+ * 2: operation (short), period type (short), adjustment (int)
+ * 3: period (int), count (int)
+ * <p>
+ * and the index when it happens stored in pureStaticCount
+ */
+public class RuntimeIntervalModelBuilder implements Mutable {
+    public static final long DYNAMIC_LO_HI = Numbers.INT_NaN + 2;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModelBuilder.java:49]: interesting choice of constant :) perhaps `Numbers.LONG_NaN` ?

File: core/src/main/java/io/questdb/std/Numbers.java [hunk 8]
```diff
@@ -514,6 +514,10 @@ public static int compare(float a, float b) {
         return Integer.compare(anotherBits, thisBits);                          // (0.0, -0.0) or (NaN, !NaN)
     }
 
+    public static short decodeHighShort(int val) {
+        return (short) (val >> (short)16);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/std/Numbers.java:518]: second cast to short is not necessary

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModel.java [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+import io.questdb.std.ObjList;
+import io.questdb.std.datetime.microtime.Timestamps;
+
+import static io.questdb.griffin.model.IntervalUtils.STATIC_LONGS_PER_DYNAMIC_INTERVAL;
+import static io.questdb.griffin.model.RuntimeIntervalModelBuilder.DYNAMIC_LO_HI;
+
+public class RuntimeIntervalModel implements RuntimeIntrinsicIntervalModel {
+    // These 2 are incoming model
+    private final LongList intervals;
+    private final ObjList<Function> dynamicRangeList;
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModel.java:40]: ok, these are closed now when factory closed

File: core/src/main/java/io/questdb/griffin/model/RuntimeIntrinsicIntervalModel.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+package io.questdb.griffin.model;
+
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.std.LongList;
+import io.questdb.std.datetime.microtime.Timestamps;
+
+public interface RuntimeIntrinsicIntervalModel {
+    LongList calculateIntervals(SqlExecutionContext sqlContext);
+    boolean isFocused(Timestamps.TimestampFloorMethod floorDd);
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/java/io/questdb/griffin/model/RuntimeIntrinsicIntervalModel.java]: ok

File: core/src/main/java/io/questdb/griffin/engine/functions/UnaryFunction.java [hunk 11]
```diff
@@ -50,4 +50,8 @@ default void toTop() {
     default void init(SymbolTableSource symbolTableSource, SqlExecutionContext executionContext) {
         getArg().init(symbolTableSource, executionContext);
     }
+
+    default boolean isRuntimeConstant() {
+        return getArg().isRuntimeConstant() || getArg().isConstant();
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/java/io/questdb/griffin/engine/functions/UnaryFunction.java]: I see constants being runtime constants but I separated them

File: core/src/main/java/io/questdb/griffin/engine/functions/BinaryFunction.java (hunk 2 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/griffin/engine/functions/BinaryFunction.java]: constant + runtime constant = runtime constant
runtime constant  + runtime constant = runtime constant
I refactored it slightly as you suggested

File: core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModel.java (hunk 4 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModel.java]: ok

File: core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModel.java (hunk 5 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModel.java]: ok

File: core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModel.java (hunk 6 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModel.java]: ok

File: core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModelBuilder.java (hunk 7 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModelBuilder.java]: I removed it, added another short flag to indicate which side of interval is the function

File: core/src/main/java/io/questdb/std/Numbers.java (hunk 8 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/std/Numbers.java]: ok

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**